### PR TITLE
Add secret provisioner mounted as file or env variable

### PR DESCRIPTION
**Pull request by mshaposhnik**
### What does this PR do?
Add ability to mount secrets as environment variable or file inso user's workspace.

### What issues does this PR fix or reference?
#14680 (partially)

#### Docs PR

https://github.com/eclipse/che-docs/pull/1287

### CODE REVIEW

**Comment by dmytro-ndp** [comment]
crw-ci-test

**Comment by dmytro-ndp** [comment]
@mshaposhnik: could you, please, update PR with latest changes from **master** branch, to avoid Happy path PR check error "_No such property: customResourceFileContent for class: WorkflowScript'_?

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/SecretAsContainerResourceProvisioner.java [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+    }
+  }
+
+  private String envName(Secret secret, String key) throws InfrastructureException {
+    String mountEnvName;
+    if (secret.getData().size() == 1) {
+      try {
+        mountEnvName =
+            firstNonNull(
+                secret.getMetadata().getAnnotations().get("envName"),
```
Comment (metlos) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/SecretAsContainerResourceProvisioner.java]: `envName` -> `org.eclipse.che/env-name` ?

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/SecretAsContainerResourceProvisioner.java [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+                secret.getMetadata().getAnnotations().get(key + ".envName"));
```
Comment (metlos) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/SecretAsContainerResourceProvisioner.java]: This may overflow the 63 chars limit on annotation name (note that the "domain name" before the `/` doesn't count towards that limit - https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/#syntax-and-character-set).

Maybe we should do something like:

`org.eclipse.che/secret_1=<name-of-secret-data>`
`org.eclipse.che/secret-env_1=<name-of-env-var-to-mount-as>`

Where the matching would be made by the string after the `_` (and it doesn't even have to be a number).

**Review by metlos — COMMENTED** [review]
A couple of mostly stylistic comments.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/SecretAsContainerResourceProvisioner.java [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+      secretLabels.put(pair[0], pair.length == 1 ? "" : pair[1]);
+    }
+  }
+
+  public void provision(E env, KubernetesNamespace namespace) throws InfrastructureException {
+    LabelSelector selector = new LabelSelectorBuilder().withMatchLabels(secretLabels).build();
+    for (Secret secret : namespace.secrets().get(selector)) {
+      boolean mountAsEnv =
+          parseBoolean(
+              secret.getMetadata().getAnnotations().getOrDefault("useSecretAsEnv", "false"));
```
[the remark this answers is not in the document]
Comment (mshaposhnik) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/SecretAsContainerResourceProvisioner.java]: They are annotations, not labels, so as for me no much reason to make them prefixed, so lets keep writer-s job easier, wdyt ?

[the remark this answers is not in the document]
Comment (mshaposhnik) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/SecretAsContainerResourceProvisioner.java]: Annotations can be random: https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties [hunk 4]
```diff
@@ -604,3 +604,8 @@ che.workspace.devfile.default_editor=eclipse/che-theia/next
 # eclipse/che-theia-exec-plugin/0.0.1,eclipse/che-theia-terminal-plugin/0.0.1,https://cdn.pluginregistry.com/vi-mode/meta.yaml
 # If the plugin is a URL, the plugin's meta.yaml is retrieved from that URL.
 che.workspace.devfile.default_editor.plugins=eclipse/che-machine-exec-plugin/nightly
+
+# Defines comma-separated list of labels for selecting secrets from a user namespace,
+# which will be mount into workspace containers as a files or env variables.
+# Only secrets that match ALL given labels will be selected.
+che.workspace.provision.secret.labels=app.kubernetes.io/part-of:org.eclipse.che,app.kubernetes.io/component: workspace-secret
```
[the remark this answers is not in the document]
Comment (mshaposhnik) [reply assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties]: Makes sense, will change to equality-based requirements format

[the remark this answers is not in the document]
Comment (mshaposhnik) [reply assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties]: Unfortunately, seems there is no native support for equality-based strings in K8S client.
So if we want to support like `a=b,c!=d` we need to parse them manually using regex an convert into Set-based requirement like `a In b, c NotIn d`. Not sure if that worth it.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/SecretAsContainerResourceProvisioner.java (hunk 2 above)
Comment (mshaposhnik) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/SecretAsContainerResourceProvisioner.java]: User will be just unable to create such secret. So i would like to leave it as is, 55 symbols is more than enough for data entry key, and that you proposed will be extremely hard to explain to user what he should write in his secrets, and, especially, correctly state that in exception message (i am not even sure i did get it right :) ).

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/SecretAsContainerResourceProvisioner.java [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+@Beta
+@Singleton
+public class SecretAsContainerResourceProvisioner<E extends KubernetesEnvironment> {
+
+  private final Map<String, String> secretLabels;
+
+  @Inject
+  public SecretAsContainerResourceProvisioner(
+      @Named("che.workspace.provision.secret.labels") String[] labels) {
+    this.secretLabels = new HashMap<>();
```
[the remark this answers is not in the document]
Comment (mshaposhnik) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/SecretAsContainerResourceProvisioner.java]: Do you find it more readable?

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/SecretAsContainerResourceProvisioner.java [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+  }
+
+  public void provision(E env, KubernetesNamespace namespace) throws InfrastructureException {
+    LabelSelector selector = new LabelSelectorBuilder().withMatchLabels(secretLabels).build();
+    for (Secret secret : namespace.secrets().get(selector)) {
+      boolean mountAsEnv =
+          parseBoolean(
+              secret.getMetadata().getAnnotations().getOrDefault("useSecretAsEnv", "false"));
+      String targetContainerName = secret.getMetadata().getAnnotations().get("targetContainer");
+      if (mountAsEnv) {
```
[the remark this answers is not in the document]
Comment (mshaposhnik) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/SecretAsContainerResourceProvisioner.java]: Looks you're not reviewving latest version

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/SecretAsContainerResourceProvisioner.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (skabashnyuk) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/SecretAsContainerResourceProvisioner.java]:
```
    secretLabels = Arrays.stream(labels)
        .map(l -> l.split("=", 2))
        .map(p -> p[1] = p.length == 1 ? "" : p[1])
        .collect(Collectors.toMap(p -> p[0], p -> p[1]));
```

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/SecretAsContainerResourceProvisioner.java (hunk 2 above)
Comment (mshaposhnik) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/SecretAsContainerResourceProvisioner.java]: So what with that @metlos @sparkoo ? I am not a big fan of creating 2 separate properties to describe single thing. it's very hard to document and extremely hard to provide correct error message in case something is missing. 55 symbols (63 - `_envName` ) is more than enougth for data key, otherwise, user will just not be able to create such secret (our code does not creates it).

Comment (metlos) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/SecretAsContainerResourceProvisioner.java]: I'm ok with explaining the 55 character limit in the docs.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/SecretAsContainerResourceProvisioner.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (metlos) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/SecretAsContainerResourceProvisioner.java]: Taken from the above mentioned docs:

> If the prefix is omitted, the annotation Key is presumed to be private to the user. Automated system components (e.g. kube-scheduler, kube-controller-manager, kube-apiserver, kubectl, or other third-party automation) which add annotations to end-user objects must specify a prefix.

So I think it is actually required of us to use the prefix. You could argue that we're just reading the annotation, not adding it as the docs suggest, but my counter-argument would be that we are the sole users of that annotation and that we should not "invade" the user's "namespace".

There is also a prior art to this style of annotation declarations - the ingress annotations that are also defined by the user, yet they are prefixed, because they are specific to certain ingress controllers.

[the remark this answers is not in the document]
Comment (mshaposhnik) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/SecretAsContainerResourceProvisioner.java]: Prefixes was already agreed with Mario and applied.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/SecretAsContainerResourceProvisioner.java [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+            firstNonNull(
+                secret.getMetadata().getAnnotations().get(ANNOTATION_ENV_NAME),
+                secret
+                    .getMetadata()
+                    .getAnnotations()
+                    .get(format(ANNOTATION_ENV_NAME_TEMPLATE, key)));
+      } catch (NullPointerException e) {
+        throw new InfrastructureException(
+            format(
+                "Unable to mount secret '%s': it has to be mounted as Env, but env name is not specified."
```
Comment (sparkoo) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/SecretAsContainerResourceProvisioner.java]: Similar as above, I would rephrase that a bit.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/SecretAsContainerResourceProvisioner.java [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+                    + "Please make sure you specified '%s' annotation of secret.",
+                secret.getMetadata().getName(), ANNOTATION_ENV_NAME));
+      }
+    } else {
+      mountEnvName =
+          secret.getMetadata().getAnnotations().get(format(ANNOTATION_ENV_NAME_TEMPLATE, key));
+      if (mountEnvName == null) {
+        throw new InfrastructureException(
+            format(
+                "Unable to mount key '%s' of secret '%s': it has to be mounted as Env, but env name is not specified."
```
Comment (sparkoo) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/SecretAsContainerResourceProvisioner.java]: same as above

File: infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/SecretAsContainerResourceProvisionerTest.java [hunk 9]
```diff
@@ -0,0 +1,350 @@
...
+import org.eclipse.che.workspace.infrastructure.kubernetes.namespace.KubernetesNamespace;
+import org.eclipse.che.workspace.infrastructure.kubernetes.namespace.KubernetesSecrets;
+import org.mockito.Mock;
+import org.mockito.testng.MockitoTestNGListener;
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.Listeners;
+import org.testng.annotations.Test;
+
+@Listeners(MockitoTestNGListener.class)
+public class SecretAsContainerResourceProvisionerTest {
```
Comment (sparkoo) [inline infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/SecretAsContainerResourceProvisionerTest.java:58]: can you please test that secret with both mountAs file and env works? Or it behaves how we expect...

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/SecretAsContainerResourceProvisioner.java [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+  }
+
+  public void provision(E env, KubernetesNamespace namespace) throws InfrastructureException {
+    LabelSelector selector = new LabelSelectorBuilder().withMatchLabels(secretLabels).build();
+    for (Secret secret : namespace.secrets().get(selector)) {
+      String targetContainerName =
+          secret.getMetadata().getAnnotations().get(ANNOTATION_TARGET_CONTAINER);
+      if ("env".equalsIgnoreCase(secret.getMetadata().getAnnotations().get(ANNOTATION_MOUNT_AS))) {
+        mountAsEnv(env, secret, targetContainerName);
+      } else {
```
Comment (metlos) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/SecretAsContainerResourceProvisioner.java]: are we ok with `.../mount-as: boo` being interpreted as "mount as file"?

IMHO, this should warn if the annotation is anything else but the supported values and only interpret the lack of the annotation as the default value - mount as file.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/SecretAsContainerResourceProvisioner.java [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+            .map(item -> item.split("=", 2))
+            .collect(toMap(p -> p[0], p -> p.length == 1 ? "" : p[1]));
+  }
+
+  public void provision(E env, KubernetesNamespace namespace) throws InfrastructureException {
+    LabelSelector selector = new LabelSelectorBuilder().withMatchLabels(secretLabels).build();
+    for (Secret secret : namespace.secrets().get(selector)) {
+      String targetContainerName =
+          secret.getMetadata().getAnnotations().get(ANNOTATION_TARGET_CONTAINER);
+      if ("env".equalsIgnoreCase(secret.getMetadata().getAnnotations().get(ANNOTATION_MOUNT_AS))) {
```
[the remark this answers is not in the document]
Comment (mshaposhnik) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/SecretAsContainerResourceProvisioner.java]: No. it can't

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/SecretAsContainerResourceProvisioner.java [hunk 12]
```diff
@@ -0,0 +1,211 @@
...
+  }
+
+  private void mountAsFile(E env, Secret secret, String targetContainerName)
+      throws InfrastructureException {
+
+    final String mountPath = secret.getMetadata().getAnnotations().get(ANNOTATION_MOUNT_PATH);
+    if (mountPath == null) {
+      throw new InfrastructureException(
+          format(
+              "Unable to mount secret '%s': it has to be mounted as file, but mountPath is not specified."
```
Comment (metlos) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/SecretAsContainerResourceProvisioner.java]: Unable to mount secret '%s': It is configured to be mounted as a file but the mount path was not specified. Please define the '%s' annotation on the secret to specify it.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/SecretAsContainerResourceProvisioner.java [hunk 13]
```diff
@@ -0,0 +1,211 @@
...
+        continue;
+      }
+      if (podData
+          .getSpec()
+          .getVolumes()
+          .stream()
+          .anyMatch(v -> v.getName().equals(volumeFromSecret.getName()))) {
+        throw new InfrastructureException(
+            format(
+                "Volume name '%s' provisioned from secret, clashes with existing volume name.",
```
Comment (metlos) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/SecretAsContainerResourceProvisioner.java]: Couldn't the volume name be actually randomized to avoid the clashes?

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/SecretAsContainerResourceProvisioner.java [hunk 14]
```diff
@@ -0,0 +1,211 @@
...
+        if (targetContainerName != null && !container.getName().equals(targetContainerName)) {
+          continue;
+        }
+        if (container
+            .getVolumeMounts()
+            .stream()
+            .anyMatch(vm -> vm.getMountPath().equals(mountPath))) {
+          throw new InfrastructureException(
+              format(
+                  "Volume Mount path '%s' provisioned from secret, clashes with existing volume mount path.",
```
Comment (metlos) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/SecretAsContainerResourceProvisioner.java]: The secret '%s' defines a mount path '%s' that clashes with another volume mount path already present on the workspace pod.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/SecretAsContainerResourceProvisioner.java [hunk 15]
```diff
@@ -0,0 +1,211 @@
...
+    if (secret.getData().size() == 1) {
+      try {
+        mountEnvName =
+            firstNonNull(
+                secret.getMetadata().getAnnotations().get(ANNOTATION_ENV_NAME),
+                secret
+                    .getMetadata()
+                    .getAnnotations()
+                    .get(format(ANNOTATION_ENV_NAME_TEMPLATE, key)));
+      } catch (NullPointerException e) {
```
Comment (metlos) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/SecretAsContainerResourceProvisioner.java:194]: Exceptions shouldn't be used for branching IMHO.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/SecretAsContainerResourceProvisioner.java (hunk 7 above)
Comment (metlos) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/SecretAsContainerResourceProvisioner.java]: Unable to mount the secret '%s'. It is configured to be mounted as an environment variable but the name of the environment variable is not specified. Please define the '%s' annotation on the secret to specify it.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/SecretAsContainerResourceProvisioner.java (hunk 8 above)
Comment (metlos) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/SecretAsContainerResourceProvisioner.java]: Unable to mount the key '%s' of the secret '%s'. It is configured to be mounted as an environment variable but the name of the environment variable is not specified. Please define the '%s' annotation on the secret to specify it.

File: infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/SecretAsContainerResourceProvisionerTest.java (hunk 9 above)
Comment (mshaposhnik) [reply infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/SecretAsContainerResourceProvisionerTest.java:58]: How can annotations have the same key twice?

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/SecretAsContainerResourceProvisioner.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (mshaposhnik) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/SecretAsContainerResourceProvisioner.java]: Made it strictrly validated.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/SecretAsContainerResourceProvisioner.java (hunk 10 above)
Comment (mshaposhnik) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/SecretAsContainerResourceProvisioner.java]: Made it strict validated

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/SecretAsContainerResourceProvisioner.java (hunk 15 above)
Comment (mshaposhnik) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/SecretAsContainerResourceProvisioner.java:194]: Yes, but it's not pretty much branching, just rethrowring another type of exc

File: infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/SecretAsContainerResourceProvisionerTest.java (hunk 9 above)
Comment (sparkoo) [reply infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/SecretAsContainerResourceProvisionerTest.java:58]: they can't I guess :)

**Review by sparkoo — APPROVED** [review]

**Review by skabashnyuk — APPROVED** [review]
